src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp

```diff
@@ -117,7 +117,6 @@ void G1FullGCCompactionPoint::forward(oop object, size_t size) {
       // Make sure object has the correct mark-word set or that it will be
       // fixed when restoring the preserved marks.
       assert(object->mark() == markWord::prototype_for_klass(object->klass()) || // Correct mark
-             object->mark_must_be_preserved() || // Will be restored by PreservedMarksSet
```

Comment:
It would seem this assert checks that mark is correct or it would be preserved. Since all marks are now preserved, this assert is basically `true` at all times? If you just drop `object->mark_must_be_preserved()`, it should be failing with forwarded objects, no?

Reply:
This block concerns about non-forwarded objects because forwarded objects are handled above.

Reply:
OK, but I am still confused what `mark_must_be_preserved` case does in the original assert. Faithfully, you would need to replace that case with `true`, and then the whole assert should fold to always passing? Maybe it handles identityHashCode-bearing or displaced-header objects?

Reply:
Right. Originally, only relatively few headers need to be preserved. Specifically, all headers that are involved in locking or which carry a hash-code. The assert checks that we either have a header that gets preserved, or a default/prototype header. Now that all headers are preserved, the asserts fold-up to be always true and can go away. I'll fix that.